Rewrite Arm Subsystems

Zoarial94 opened the pull request.

Zoarial94 committed:
Created RangedPiecewise.kt and 1 unit test

This class will be used in an upcoming super-system used to manage multiple subsystem.
This class will guarantee there are no gaps in the input which will not calculate a valid output and more.

Zoarial94 committed:
Moved Kotlin files and made RangedPiecewise.kt use generics

Added a few more test.

Zoarial94 committed:
Moved Piecewise files into the correct source directory

Zoarial94 committed:
RangedPiecewise will now verify on construction

The constructor will throw if the range can't be verified.

Zoarial94 committed:
Linting cleanup

Zoarial94 committed:
Merge branch 'main' into Zoarial94-Ranged-Piecewise

Zoarial94 committed:
Linting cleanup. Fixed file name.

Zoarial94 committed:
Linting cleanup.

Zoarial94 committed:
Added some documentation to the Piecewise classes

Zoarial94 committed:
Fixed linter issues

GearBoxFox committed:
Feat: Refactored Swerve to no longer use SwerveIO

GearBoxFox committed:
Feat: Refactored Swerve teleopdrive.java to no longer use IO files

GearBoxFox committed:
Feat: Refactored Wrist to no longer use IO files

Grandcircuit424 committed:
FEAT: Started working on the arm supersystem

Grandcircuit424 committed:
Merge remote-tracking branch 'origin/RewriteSubsystems' into RewriteSubsystems

GearBoxFox committed:
Feat: Refactored Arm to no longer use IO files

GearBoxFox committed:
Merge remote-tracking branch 'origin/RewriteSubsystems' into RewriteSubsystems

Zoarial94 committed:
Updated AutoRetry, .gitignore, and vendordeps

Added certain simulation files to .gitignore

Added GirlsOfSteel and SnobotSim to vendordeps

Zoarial94 committed:
Enabled simulation support and fixed logging directory

GearBoxFox committed:
Feat: created a factory to create spark max motor controller

GearBoxFox committed:
Fix: minor style guide fixes

Zoarial94 committed:
Merge branch 'RewriteSubsystems' of github.com:TitaniumTitans/2023ChargedUp into RewriteSubsystems

Zoarial94 committed:
Fixed reporting AutoRetry errors. Minor cleanup.

Renamed SwerveModule package

DriverStation will only show an error once. Added the object hashcode to the error string to force it to output.

GearBoxFox committed:
Fix: SuperSystem no longer looks for one arm subsystem

GearBoxFox committed:
Merge remote-tracking branch 'origin/RewriteSubsystems' into RewriteSubsystems

# Conflicts:
#	src/main/java/frc/robot/supersystems/ArmSupersystem.java

Grandcircuit424 committed:
FIX: Adjusting default parameters, Also added autoRetry to parameters

GearBoxFox committed:
Feat: implemented limiting logic for supersystem

GearBoxFox committed:
Merge remote-tracking branch 'origin/RewriteSubsystems' into RewriteSubsystems

Grandcircuit424 committed:
FIX: Move factory function to companion

Steggoo committed:
Feat: Added new arm angle debugging tab.

Steggoo committed:
Merge branch 'RewriteSubsystems' of https://github.com/TitaniumTitans/2023ChargedUp into RewriteSubsystems

Grandcircuit424 committed:
FIX: SparkMAXs in subsystems are now made with factory.

Grandcircuit424 committed:
Merged upstream

Steggoo committed:
Feat: Added dashboard variables for arm extension

Steggoo committed:
Merge branch 'RewriteSubsystems' of https://github.com/TitaniumTitans/2023ChargedUp into RewriteSubsystems

GearBoxFox committed:
Feat: added command for controlling supersystem

GearBoxFox committed:
Merge remote-tracking branch 'origin/RewriteSubsystems' into RewriteSubsystems

Grandcircuit424 committed:
Fix: Switched File Name

Steggoo committed:
Merge branch 'RewriteSubsystems' of https://github.com/TitaniumTitans/2023ChargedUp into RewriteSubsystems

GearBoxFox committed:
Fix: removed unused imports

GearBoxFox committed:
Merge remote-tracking branch 'origin/RewriteSubsystems' into RewriteSubsystems

Steggoo committed:
Merge remote-tracking branch 'origin/RewriteSubsystems' into RewriteSubsystems

Zoarial94 committed:
Removed unused commands

Steggoo committed:
Feat: Added dashboard variables for wrist subsystem

Steggoo committed:
Merge branch 'RewriteSubsystems' of https://github.com/TitaniumTitans/2023ChargedUp into RewriteSubsystems

GearBoxFox committed:
Feat: created testing limits for supersystem

GearBoxFox committed:
Merge remote-tracking branch 'origin/RewriteSubsystems' into RewriteSubsystems

Grandcircuit424 committed:
Feat: Added function to reinitialize spark CAN frame periods

GearBoxFox committed:
Fix: restructured shuffleboard logging to not crash robot

GearBoxFox committed:
Merge remote-tracking branch 'origin/RewriteSubsystems' into RewriteSubsystems

Zoarial94 reviewed (commented):
- Constants.java
  - Piecewise limits should be GOSProperties to allow for dynamic updates

- RobotContainer.java
  - The Commands Shuffleboard is not being populated with commands. Just pose values.

- SupersystemToPoseCommand
  - withInterruptBehaviour is not used correctly

- ArmAngleSubsystem
   - Arm angle limits will be ~30 for stow and ~300 for full rotation. Make sure limit code is updated since these values will be flipped from the inital code.

- ArmExtSubsystem
  - Why does the extension use an external PID controller?

Still reviewing.

Zoarial94 committed:
GearboxFox Changes

Zoarial94 committed:
Upgraded WPILib

Zoarial94 commented on src/main/java/frc/robot/subsystems/arm/ArmAngleSubsystem.java:
```diff
@@ -0,0 +1,139 @@
+    }
+
+    public double getArmAngle() {
+        return Utils.normalize((m_encoderArmAngle.getAbsolutePosition() * 360) - ArmConstants.ARM_OFFSET);
```
Refactor to return ~30 degrees for stow, and ~300 degrees for intake

GearBoxFox replied:
We'll have to do this once the robot is toghether, as the offset is hardware specific

Zoarial94 commented on src/main/java/frc/robot/subsystems/arm/ArmExtSubsystem.java:
```diff
@@ -0,0 +1,116 @@
+        m_relativeEncoderArmEx = m_ArmEx.getEncoder();
+        m_relativeEncoderArmEx.setPositionConversionFactor(Constants.ArmConstants.EXTENSION_RATIO);
+
+        m_extPID = new PIDController(Constants.ArmConstants.ARM_EXT_KP, Constants.ArmConstants.ARM_EXT_KI, Constants.ArmConstants.ARM_EXT_KD);
```
Refactor to use the motor controller's PID controller

Zoarial94 commented on src/main/java/frc/robot/subsystems/arm/ArmExtSubsystem.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    public void setArmSpeed(double speed) {
+        if(armAtLowerLimit() && speed <= 0){
```
Added extension soft limit to prevent over-extending

GearBoxFox committed:
Fix: testCommand tab now adds proper commands

Zoarial94 commented on src/main/java/frc/robot/subsystems/wrist/WristSubsystem.java:
```diff
@@ -1,73 +1,221 @@
+        m_tofSensor = new TimeOfFlight(WristConstants.TOF_PORT);
+        m_tofSensor.setRangingMode(RangingMode.Short, 10);
+
+        m_wristPID = new PIDController(WristConstants.WRIST_KP, WristConstants.WRIST_KI, WristConstants.WRIST_KD);
```
Refactor to use internal PID controller

GearBoxFox committed:
Feat: subsystem use constants from LimitsConstants everywhere

Zoarial94 commented on src/main/java/frc/robot/subsystems/wrist/WristSubsystem.java:
```diff
@@ -1,73 +1,221 @@
+
+        m_wristZeroLimit = new DigitalInput(WristConstants.LIMIT_SWITCH_PORT);
+
+        m_tofSensor = new TimeOfFlight(WristConstants.TOF_PORT);
```
Look into reliability of the sensor data. If the sensor is reading the intake itself, the FOV can be adjusted

Zoarial94 commented on src/main/kotlin/lib/factories/SparkMaxFactory.kt:
```diff
@@ -0,0 +1,97 @@
+        const val MAX_CAN_FRAME_PERIOD = 65535;
+
+        val listOfAllSparksAndConfigs:MutableList<SparkWithConfig> = ArrayList();
+        fun updateCanFramePeriods() {
```
Create a periodic task in the robot to run this check occasionally.

Zoarial94 commented on src/main/java/frc/robot/supersystems/ArmLimits.java:
```diff
@@ -0,0 +1,45 @@
+
+    public boolean withinWristLimit(double inside) {
+        return (
+                (wristRange.getLeft() > inside || (wristRange.getLeftInclusive() && wristRange.getLeft() == inside)) ||
```
Refactor and put these calculations in the Range data class

Zoarial94 commented on src/main/java/frc/robot/supersystems/ArmPose.java:
```diff
@@ -0,0 +1,16 @@
+
+    public ArmPose(double ext, double angle, double wrist) {
+        if (ext < 0.0 || angle < 0.0 || wrist < 0.0) {
+            throw new RuntimeException("Not a valid ArmPose parameter");
```
Don't throw. Clamp the value. Don't want to robot to unexpectedly stop while running for a minor issue

GearBoxFox committed:
Feat: changed LimitConstants to use GosDoubleProperties

Zoarial94 committed:
Made ArmPose Sendable.

Zoarial94 committed:
Set higher currently limit for SwerveModNeo and minor cleanup

GearBoxFox committed:
Feat: ArmAngleSubsystem.getArmAngle now returns 0 as stow position

GearBoxFox committed:
Merge remote-tracking branch 'origin/RewriteSubsystems' into RewriteSubsystems

Steggoo committed:
Fix: Changed arm angle and extension to display data properly.

Steggoo committed:
Merge branch 'RewriteSubsystems' of https://github.com/TitaniumTitans/2023ChargedUp into RewriteSubsystems

# Conflicts:
#	src/main/java/frc/robot/subsystems/arm/ArmAngleSubsystem.java
#	src/main/java/frc/robot/subsystems/arm/ArmExtSubsystem.java

GearBoxFox committed:
Feat: toggle arm angle brake mode from dashboard

Steggoo committed:
Merge branch 'RewriteSubsystems' of https://github.com/TitaniumTitans/2023ChargedUp into RewriteSubsystems

Zoarial94 committed:
Revert "Upgraded WPILib"

This reverts commit e007b855f12b3a0acb1c92b190ac76c9c5d1895b.

Steggoo committed:
Fix: Changed dashboard name for the raw encoder angle.

Zoarial94 commented:
https://github.com/REVrobotics/SPARK-MAX-Examples/blob/master/Java/Position%20Closed%20Loop%20Control/src/main/java/frc/robot/Robot.java

GearBoxFox committed:
Updated .gitignore

GearBoxFox committed:
Removed simgui-ds.json and .vscode

Steggoo committed:
Feat: Changed arm extension PID controller to the integrated PID controller.

GearBoxFox committed:
Merge remote-tracking branch 'origin/RewriteSubsystems' into RewriteSubsystems

Steggoo committed:
Feat: Added motor output to arm angle, wrist angle, and intake.

Steggoo committed:
Merge branch 'RewriteSubsystems' of https://github.com/TitaniumTitans/2023ChargedUp into RewriteSubsystems

GearBoxFox committed:
Feat: spark max factory sets frame 0 to 100 milliseconds

GearBoxFox committed:
Merge branch 'RewriteSubsystems' of https://github.com/TitaniumTitans/2023ChargedUp into RewriteSubsystems

Zoarial94 committed:
Added auto logging to ArmExtSubsystem

Zoarial94 committed:
Adjustments to the Commands Shuffleboard tab

GearBoxFox committed:
Feat: created foot pedal custom controller

GearBoxFox committed:
Button bound foot pedals

GearBoxFox committed:
Fix: updated AdvantageKit to wpilib version 2023.4.1

Zoarial94 committed:
Updated wrist current limit based off of testing

GearBoxFox committed:
Use default config with CreateSparkMax

GearBoxFox committed:
Merge branch 'RewriteSubsystems' of https://github.com/TitaniumTitans/2023ChargedUp into RewriteSubsystems

GearBoxFox committed:
Checked arm angle values

GearBoxFox committed:
Changed max angle speed and test poses

GearBoxFox committed:
default commands

GearBoxFox committed:
Fix: soft limits for manual control

Zoarial94 committed:
Removed some shuffleboard data to lower periodic loop time. Adjusted pose setpoint.

Zoarial94 committed:
Added new test and clamped the value going into PiecewiseInterval

Zoarial94 committed:
Fixed setAngleSpeed

Zoarial94 committed:
Fixed RangedPiecewise crashing when calculating outside range.
Fixed ArmAngleSubsystem setAngleSpeedFactory to keep running output, otherwise the default command will set power back to 0.

GearBoxFox committed:
Fix: debugged some limits, inversions, and manual controls with arm supersystem

GearBoxFox committed:
Removed some uneeded logging for sonarqube

GearBoxFox committed:
Make sure that wrist pivot motor is in brake mode

GearBoxFox committed:
Feat: periodically check for reset motor and reapply CAN configs on it

GearBoxFox committed:
Fix: spaced timer out and reseting timestamp to stop from only calling once

GearBoxFox committed:
Use a non-static timer object

GearBoxFox committed:
Feat: rerun all configs on robot enable and hope that this doesn't cause issues with delays

GearBoxFox committed:
Fix: added withinRange() function to range data class

Zoarial94 commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -23,6 +25,7 @@ public class Robot extends LoggedRobot {
   private Command m_autonomousCommand;
   private RobotContainer m_robotContainer;
 
+  private Timer m_timer;
```
Timer needs to be initialized. Other than that, I believe the code is ready to be merged into main.

GearBoxFox replied:
This has been fixed, I'm pulling into main now.

GearBoxFox committed:
Fix: initialzed m_timer

GearBoxFox committed:
various limiting changes and pose changes for testing intake and scoring

GearBoxFox committed:
wrote down more setpoints for arm scoring

GearBoxFox committed:
Finished finding arm setpoints

GearBoxFox committed:
Added an extra setpoint to the dashboard

GearBoxFox committed:
Dashboard setpoints for all used poses and increased arm angle speed

GearBoxFox committed:
Feat: increased arm ext max limit

GearBoxFox reviewed (approved):
Code has been tested on physical robot and is ready to be pulled into main for continuing development.
